frontend/src/Connection.ts

```diff
@@ -565,3 +588,29 @@ export class Connection {
     this.bindSocket();
   };
 }
+
+/**
+ * Fire a function if the timer runs out. You can reset it, or
+ * cancel it to prevent the function from being fired.
+ *
+ * @param onExpired
+ * @param timeoutMs
+ * @returns
+ */
+function resettableTimeout(
+  onExpired: () => void,
+  timeoutMs: Types.Milliseconds
+) {
+  let timer = setTimeout(onExpired, timeoutMs);
+
+  return {
+    reset() {
+      clearTimeout(timer);
+      timer = setTimeout(onExpired, timeoutMs);
+    },
+    cancel() {
+      clearTimeout(timer);
+    },
+  };
+}
+type ResettableTimeout = ReturnType<typeof resettableTimeout>;
```

Comment:
Not perfectly clear on the format of the frontend, but is this the proper place to store the type?

Reply:
I think the types that are shoved in a separate file are mainly just the types of "incoming" data from the backend that we want to know about (or general helper types), so it feels like a good enough place to me offhand :)